src/vsphere_cpi/lib/cloud/vsphere/attach_tag_to_vm.rb

```diff
@@ -0,0 +1,243 @@
+      end
+
+      def to_s
+        "Duplicated tags found in tag category '#{@category_name}', deduplicated and continue."
+      end
+    end
+
+    class CreateTagHashCategoryError < StandardError
+      def initialize(type)
+        @type = type
+      end
+
+      def to_s
+        "#{@type} in cloud config , skip processing this category-tag pair."
+      end
+    end
+
+    class CreateTagHashTagError < StandardError
+      def initialize(type, category_name)
+        @type = type
+        @category_name = category_name
+      end
+
+      def to_s
+        "#{@type} in category '#{@category_name}', skip attaching this tag."
+      end
+    end
+
+    class AttachTagToVm
+      include Logger
+      def self.InitializeConnection(cloud_config, logger)
+        configuration = VSphereAutomation::Configuration.new.tap do |config|
+          config.host = cloud_config.vcenter_host
+          config.username = cloud_config.vcenter_user
+          config.password = cloud_config.vcenter_password
+          config.scheme = 'https'
+          config.verify_ssl = false
+          config.verify_ssl_host = nil
+          config.logger = logger
+          config.debugging = false
+        end
+        api_client = VSphereAutomation::ApiClient.new(configuration)
+        api_client.default_headers['Authorization'] = configuration.basic_auth_token
+        session_api = VSphereAutomation::CIS::SessionApi.new(api_client)
+        session_id = session_api.create('').value
+        api_client.default_headers['vmware-api-session-id'] = session_id
+        api_client
+      end
+
+      def initialize(api_client)
+        @api_client = api_client
+      end
+
+      private def tag_association_api
+        @tag_association_api ||= VSphereAutomation::CIS::TaggingTagAssociationApi.new(@api_client)
+      end
+
+      private def tagging_tag_api
+        @tagging_tag_api ||= VSphereAutomation::CIS::TaggingTagApi.new(@api_client)
+      end
+
+      private def tagging_category_api
+        @tagging_category_api ||= VSphereAutomation::CIS::TaggingCategoryApi.new(@api_client)
+      end
+
+      def retrieve_category_id(category_name, category_ids)
+        category_ids.find { |category_id| tagging_category_api.get(category_id).value.name == category_name }
+      end
+
+      def retrieve_tag_id(tag_name, tag_id_list)
+        tag_id_list.find { |tag_id| tagging_tag_api.get(tag_id).value.name == tag_name }
+      end
+
+      def vm_association?(target_category_id)
+        category_information = tagging_category_api.get(target_category_id)
+        category_assoc_types = category_information.value.associable_types
+        category_assoc_types.empty? || category_assoc_types.include?("VirtualMachine")
+      end
+
+      def create_tag_hash(vm_config_tags)
+        tag_hash = Hash.new
+        vm_config_tags.each do |vm_config_tag|
+          begin
+            raise CreateTagHashCategoryError.new("Missing category content") if vm_config_tag["category"].nil?
```

Comment:
Can we name this to `BadCategoryTagInfoError`. The reason is that CreateTagHashCategoryError focuses on the function called but the `BadCategoryTagInfoError`  on what error is there.

Reply:
Done.